src/lib.rs

```diff
@@ -1,19 +1,107 @@
+/*!
+  # Selecting the right chip
+
+    This crate requires you to specify your target chip as a feature.
+
+    Please select one of the following
+
+    (Note: `x` denotes any character in [a-z])
+    *   stm32f301xb
+    *   stm32f301xc
+    *   stm32f301xd
+    *   stm32f301xe
+    *   stm32f318
+    *   stm32f302
+    *   stm32f303xb
+    *   stm32f303xc
+    *   stm32f303xd
+    *   stm32f303xe
+    *   stm32f303x6
+    *   stm32f303x8
+    *   stm32f373
+    *   stm32f378
+    *   stm32f334
+    *   stm32f328
+    *   stm32f358
+    *   stm32f398
+
+    Example: The STM32F3Discovery board has a STM32F303VCT6 chip.
+    So you want to expand your call to `cargo` with `--features stm32f303xc`.
+
+    For more information, see the [README](https://github.com/stm32-rs/stm32f3xx-hal/blob/master/README.md#selecting-the-right-chip)
+ */
 #![no_std]
 #![allow(non_camel_case_types)]
 
-#[cfg(not(feature = "device-selected"))]
+#[cfg(all(not(feature = "device-selected"), not(feature = "needs-subvariant")))]
 compile_error!(
-    "This crate requires one of the following device features enabled:
-        stm32f301
-        stm32f318
-        stm32f302
-        stm32f303
-        stm32f373
-        stm32f378
-        stm32f334
-        stm32f328
-        stm32f358
-        stm32f398"
+    "This crate requires you to specify your target chip as a feature.
+
+    Please select one of the following
+
+    (Note: `x` denotes any character in [a-z])
+    *   stm32f301xb
+    *   stm32f301xc
+    *   stm32f301xd
+    *   stm32f301xe
+    *   stm32f318
+    *   stm32f302
+    *   stm32f303xb
+    *   stm32f303xc
+    *   stm32f303xd
+    *   stm32f303xe
+    *   stm32f303x6
+    *   stm32f303x8
+    *   stm32f373
+    *   stm32f378
+    *   stm32f334
+    *   stm32f328
+    *   stm32f358
+    *   stm32f398
+
+    Example: The STM32F3Discovery board has a STM32F303VCT6 chip.
+    So you want to expand your call to `cargo` with `--features stm32f303xc`.
+
+    For more information, see README -> Selecting the right feature gate.
+    "
+);
+
+#[cfg(all(not(feature = "device-selected"), feature = "needs-subvariant"))]
```

Comment:
I still feel like this is to much duplication, when it only adds the extra information, that some features, specifically `stm32f303` will be removed. 

I would suggest changing this to: 

```rust
#[cfg(any(feature = "stm32f303", feature = "... other devices, which are no handled (this would be stm32f302 e.g."))]
compile_error!(
        "stm32f303 feature has been removed. Choose either stm32f303xc, ...xd, ... x6, ... x8"
);
```

And revert the change on the `cfg` above. 

As no other devices are affected yet, `any` is not needed for know. But soon https://github.com/stm32-rs/stm32f3xx-hal/pull/67 will add a reason to remove `stm32f302`. 

With this change, we can think about, if we even need `needs_subvariant`. But we don't have to remove it. We could also do:

```rust
#[cfg(feature = "needs-subvariant")]
compile_error!(
        "stm32f303 feature has been removed. Choose either stm32f303xc, ...xd, ... x6, ... x8"
);
```

Reply:
I agree on the shorter error message. The rest depends on how we want to set up feature gates. maybe it's more efficient if we chat about it and come to a conclusion during that discussion.

Reply:
I like the more specific error messages introduced in https://github.com/stm32-rs/stm32f3xx-hal/pull/75/commits/0fa4b138f50ea1610355066956c2b494b918aefb  more. hope I didn't go into the completely wrong direction with this ;)